docs/tokencache.md

```diff
@@ -2,16 +2,350 @@
+
+### AccountRecord
+
+"Record" is appended in this case to avoid confusion with the public API account object.  This is where you will find all of the information about an account.  This is the object that will be serialized to storage as an Account.
+
+### Credential
+
+This is an abstract base class for all types of credentials stored in the universal cache.  This includes common fields that exist on all types of credentials.  
+
+### AccessTokenRecord
+
+Inherits from Credential and contains additional fields specific to access tokens.
+
+### RefreshTokenRecord
+
+Inherits from Credential and contains additioonal fields specific to refresh tokens.
+
+### PrimaryRefreshTokenRecord
+
+Inherits from Credential and contains additional fields specific to the primary refresh token.
+
+### IdTokenRecord
+
+Inherits from Credential and contains additional fields specific to id tokens.  
+
+> NOTE: An id token can be validated to authenticate a user.  Generally MSAL is an oAuth library and primarily concerned with authorization and access tokens; however because we perform an id token request along with every access token request.  The id token is also a credential.
+
+### Cache Key/Value Generation
+
+In order to store accounts and credentials in shared preferences first an identifier for the account and credential must be generated and the serialization of the account attributes/properties must be performed.  This is currently handled by the CacheKeyValueDelegate class which implements ICacheKeyValueDelegate interface.
+
+```java
+public interface ICacheKeyValueDelegate {
+    /**
+     * Generate cache key for a specific account.
+     *
+     * @param account Account
+     * @return String
+     */
+    String generateCacheKey(final AccountRecord account);
+
+    /**
+     * Generate cache value for a specific account.
+     *
+     * @param account Account
+     * @return String
+     */
+    String generateCacheValue(final AccountRecord account);
+
+    /**
+     * Generate cache key from the credential.
+     *
+     * @param credential Credential
+     * @return String
+     */
+    String generateCacheKey(final Credential credential);
+
+    /**
+     * Generate cache value from the credential.
+     *
+     * @param credential Credential
+     * @return String
+     */
+    String generateCacheValue(final Credential credential);
+
+    /**
+     * Get the account credential from cache value.
+     *
+     * @param string String
+     * @param t      AccountCredentialBase
+     * @param <T>    Generic type
+     * @return AccountCredentialBase
+     */
+    <T extends AccountCredentialBase> T fromCacheValue(final String string, Class<? extends AccountCredentialBase> t); // TODO consider throwing an Exception if parsing fails
+```
+
+## Storage
+
+Accounts and credentials are serialized and stored using the lowest common denominator storage method available on Andorid... SharedPreferences.  
```

Comment:
I mean, we *could* just use files, and then wrap them using BDB or whatever k-v implementation we would like.  There may be good reasons to do this...

Reply:
Not commenting on how to make this better... i think the answer is likely move to SQL Lite rather than using shared preferences... ultimately shared preference is a file... wrapped.